server/src/main/java/org/opensearch/common/util/MovingAverage.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.common.util;
+
+/**
+ * MovingAverage is used to calculate the moving average of last 'n' observations.
+ *
+ * @opensearch.internal
+ */
+public class MovingAverage {
+    private final int windowSize;
+    private final long[] observations;
+
+    private long count = 0;
+    private long sum = 0;
+    private double average = 0;
+
+    public MovingAverage(int windowSize) {
+        if (windowSize <= 0) {
+            throw new IllegalArgumentException("window size must be greater than zero");
+        }
+
+        this.windowSize = windowSize;
+        this.observations = new long[windowSize];
+    }
+
+    /**
+     * Records a new observation and evicts the n-th last observation.
+     */
+    public synchronized double record(long value) {
+        long delta = value - observations[(int) (count % observations.length)];
+        observations[(int) (count % observations.length)] = value;
+
+        count++;
+        sum += delta;
+        average = (double) sum / Math.min(count, observations.length);
+        return average;
+    }
```

Comment:
Using CAS backed by a queue or a ring-buffer, it may be possible to track the moving average similar to how it's done in indexing backpressure (though the current implementation has subtle race-condition bugs which I have highlighted in an above comment).

To successfully implement this, we need to ensure the queue size, running sum, and count of inserted elements are updated atomically; which may not be possible with CAS alone. We need to treat the entire state (even the backing queue/buffer) as immutable and create a new copy with every update (similar concept to `CopyOnWriteArrayList`).

Our use-case is write heavy (on task completion) with infrequent reads (on search backpressure service iteration), creating copies may be very expensive especially for larger window sizes. I'm still inclined to use the current approach.

Reply:
Benchmark results comparing the existing approach v/s CAS backed by an immutable ring-buffer.

Using `synchronized` (current approach):
| | 10 observations | 100 observations | 1000 observations |
| - | - | - | - |
| 1 thread | 14.123 ns/op | 14.258 ns/op | 14.544 ns/op |
| 4 threads | 368.087 ns/op | 364.879 ns/op | 378.378 ns/op |
| 16 threads | 1400.703 ns/op | 1506.456 ns/op | 1809.835 ns/op |

Using compare-and-set backed by an immutable ring-buffer ([implementation reference](https://gist.github.com/ketanv3/e6baa66cbea0f43ce6caf27989ae0e7e)):
| | 10 observations | 100 observations | 1000 observations |
| - | - | - | - |
| 1 thread | 27.438 ns/op | 112.920 ns/op | 1082.057 ns/op |
| 4 threads | 798.966 ns/op | 1077.040 ns/op | 4675.375 ns/op |
| 16 threads | 5820.276 ns/op | 8374.579 ns/op | 36474.605 ns/op |

Key observations:
* Synchronized approach doesn't have to clone the backing buffer, so, the time complexity doesn't grow proportional to the window size. This is a major drawback with the CAS approach as repeated array copy is expensive and creates a lot of GC overhead too.
* Even with smaller window sizes, the performance of CAS becomes poorer as the number of threads (and the contention) grows.
* Synchronized approach performed between 2–20x better for these measurements.